resharper/resharper-unity/src/HlslSupport/CgIncludeDirectoryTracker.cs

```diff
@@ -0,0 +1,69 @@
+using JetBrains.Application.Threading;
+using JetBrains.Application.Threading.Tasks;
+using JetBrains.Collections.Viewable;
+using JetBrains.Lifetimes;
+using JetBrains.ProjectModel;
+using JetBrains.ProjectModel.Caches;
+using JetBrains.ProjectModel.Tasks;
+using JetBrains.ReSharper.Feature.Services.Cpp.Caches;
+using JetBrains.ReSharper.Plugins.Unity.ProjectModel;
+using JetBrains.Util;
+using JetBrains.Util.Threading.Tasks;
+
+namespace JetBrains.ReSharper.Plugins.Unity.HlslSupport
+{
+    [SolutionComponent]
+    public class CgIncludeDirectoryTracker
+    {
+        private const string CG_FOLDER_PATH = "CgIncludeFolderPath";
+
+        public CgIncludeDirectoryTracker(Lifetime lifetime, UnityReferencesTracker unityReferencesTracker,
+            SolutionCaches solutionCaches, IShellLocks shellLocks, ISolutionLoadTasksScheduler scheduler,
+            CppGlobalCacheImpl cppGlobalCache, UnityVersion unityVersion, ILogger logger)
+        {
+            scheduler.EnqueueTask(new SolutionLoadTask("InitCgIncludeDirectoryTracker", SolutionLoadTaskKinds.PreparePsiModules,
+                () =>
+                {
+                    unityReferencesTracker.HasUnityReference.AdviseOnce(lifetime, _ =>
+                    {
+                        if (solutionCaches.PersistentProperties.TryGetValue(CG_FOLDER_PATH, out var result))
+                        {
+                            var oldPath = FileSystemPath.TryParse(result, FileSystemPathInternStrategy.INTERN);
+                            var newPath = GetCgIncludeFolderPath(unityVersion);
+                            if (!oldPath.Equals(newPath))
+                            {
+                                cppGlobalCache.IsInitialUpdateFinished.Change.Advise(lifetime, v =>
```

Comment:
Does this have to wait until the update is finished before invalidating? Can we cancel the update and invalidate early?

Reply:
I have changed to `IsCacheStarted`, it should be better